File: 프로그래머스/2/12978. 배달/배달.py
import heapq

def solution(N, road, K):
    graph = [[] for _ in range(N + 1)]
    distance = [int(10e8)] * (N + 1)
    
    for u, v, w in road:
        graph[u].append([v, w])
        graph[v].append([u, w])
    
    q = []
    distance[1] = 0
    heapq.heappush(q, (0, 1))
    
    while q:
        dist, now = heapq.heappop(q)
        if dist > distance[now]:
            continue 
        for node in graph[now]:
            d = dist + node[1]
            if distance[node[0]] > d:
                distance[node[0]] = d
                heapq.heappush(q, (d, node[0]))
    
	
    answer = 0
    for i in distance:
        if i <= K: answer += 1
                
    return answer

File: 프로그래머스/2/12978. 배달/배달.js
function solution(N, road, K) {
    graph = [];
    for (let i = -1; i < N; i++) graph.push([])
    for (const [u, v, w] of road) {
        graph[u].push({to:v, dist:w})
        graph[v].push({to:u, dist:w})
    }
    
    const queue = [{to:1, dist:0}];
    const dist = Array(graph.length).fill(Infinity);
    
    dist[1] = 0;
    
    while (queue.length) {
        const { to } = queue.pop();
        
        graph[to].forEach((next) => {
            const acc = dist[to] + next.dist;
            if (dist[next.to] > acc) {
                dist[next.to] = acc;
                queue.push(next);
            }
        })
    }
    
    answer = 0
    for (let i = 0; i < dist.length; i ++) {
        if (dist[i] <= K) answer++
    }
    return answer;
}